main/cron/course_announcement.php

```diff
@@ -0,0 +1,58 @@
+<?php
+/* For licensing terms, see /license.txt */
+
+require __DIR__.'/../inc/global.inc.php';
+
+if (php_sapi_name() != 'cli') {
+    exit; //do not run from browser
+}
+
+if (!api_get_configuration_value('course_announcement_scheduled_by_date')) {
+    exit;
+}
+
+$dql = "SELECT a
+            FROM ChamiloCourseBundle:CAnnouncement a
+            JOIN ChamiloCourseBundle:CItemProperty ip
+            WITH a.id = ip.ref AND a.cId = ip.course
+            WHERE
+                (a.emailSent != 1 OR
+                a.emailSent IS NULL) AND
+                ip.tool = 'announcement' AND
+                ip.visibility = 1
+            ORDER BY a.displayOrder DESC";
+
+$qb = Database::getManager()->createQuery($dql);
+$result = $qb->execute();
+
+if (!$result) {
+    exit;
+}
+
+$extraFieldValue = new ExtraFieldValue('course_announcement');
+$today = date('Y-m-d');
+
+foreach ($result as $announcement) {
+
+    $sendNotification = $extraFieldValue->get_values_by_handler_and_field_variable($announcement->getId(), 'send_notification_at_a_specific_date');
+
+    if ($sendNotification['value'] == 1) {
+
+        $dateToSend = $extraFieldValue->get_values_by_handler_and_field_variable($announcement->getId(), 'date_to_send_notification');
+
+        if ($today >= $dateToSend['value']) {
+            $courseInfo = api_get_course_info_by_id($announcement->getCId());
+
+            $query = "SELECT ip FROM ChamiloCourseBundle:CItemProperty ip
+                        WHERE ip.ref = :announcementId AND ip.course = :courseId";
+
+            $sql = Database::getManager()->createQuery($query);
+            $itemProperty = $sql->execute(['announcementId' => $announcement->getId(), 'courseId' => $announcement->getCId()]);
+            $senderId = $itemProperty[0]->getInsertUser()->getId();
+            $sendToUsersInSession = $itemProperty[0]->getSession() ? $itemProperty[0]->getSession()->getId() : false;
```

Comment:
Probar solamente $itemProperty[0]->getSession() no funciona.
El parametro SendToUsersInSession que se pasa a la function AnnouncementEmail::send() corresponde a si el usuario que ha creado el anuncio dio check or no al parametro "Send to users in all sessions of this course". Y si esta a true entonces la fonction send() va a enviar el mensaje a todos los usuarios de todas las sesiones de este curso.
Cuando probas $itemProperty[0]->getSession() esto contiene null para el primer registro de un anuncio si ha sido creado en el curso base. Y $itemProperty[0]->getSession() van a contener el id de la sesion si el anuncio a sido creado en la sesion. Entonces no coresponde a lo que quieremos aqui.
Para saber que la opcion "Send to users in all sessions of this course" habia sido marcada, se tiene que encontrar que $itemProperty[0]->getSession() === NULL pero tambien que hay mas de un solo registro y que en uno de estos otros registros de itemProperty hay uno que contiene un valor en sessionId.

Cuando se marca la opcion "Send to users in all sessions of this course" entonces en la base de datos tenemos algo asi : 

```
MariaDB [testing4]> select * from c_announcement where iid = 20;
+-----+------+------+----------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------+------------+---------------+------------+------------+
| iid | id   | c_id | title                                                                                              | content                                                                                                   | end_date   | display_order | email_sent | session_id |
+-----+------+------+----------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------+------------+---------------+------------+------------+
|  20 |   20 |   12 | Anuncio con fecha de envio activado a la fecha de antes de ayer nuevo y envio a todas las sesiones | <p>Anuncio con fecha de envio activado a la fecha de antes de ayer nuevo y envio a todas las sesiones</p> | 2021-07-21 |             9 |       NULL |          0 |
+-----+------+------+----------------------------------------------------------------------------------------------------+-----------------------------------------------------------------------------------------------------------+------------+---------------+------------+------------+
1 row in set (0.001 sec)

MariaDB [testing4]> select * from c_item_property where tool = "announcement" and ref = 20 and c_id = 12;
+------+------+-------------+------------+----------------+------------+------+--------------+---------------------+---------------------+-----+---------------------+------------------+------------+---------------+-------------+
| iid  | c_id | to_group_id | to_user_id | insert_user_id | session_id | id   | tool         | insert_date         | lastedit_date       | ref | lastedit_type       | lastedit_user_id | visibility | start_visible | end_visible |
+------+------+-------------+------------+----------------+------------+------+--------------+---------------------+---------------------+-----+---------------------+------------------+------------+---------------+-------------+
| 2609 |   12 |        NULL |         61 |             91 |       NULL | 2609 | announcement | 2021-07-21 08:07:13 | 2021-07-21 08:07:13 |  20 | AnnouncementAdded   |               91 |          1 | NULL          | NULL        |
| 2610 |   12 |        NULL |         61 |             91 |          1 | 2610 | announcement | 2021-07-21 08:07:13 | 2021-07-21 08:07:13 |  20 | AnnouncementUpdated |               91 |          1 | NULL          | NULL        |
| 2611 |   12 |        NULL |         43 |             91 |          1 | 2611 | announcement | 2021-07-21 08:07:13 | 2021-07-21 08:07:13 |  20 | AnnouncementUpdated |               91 |          1 | NULL          | NULL        |
| 2612 |   12 |        NULL |         20 |             91 |          1 | 2612 | announcement | 2021-07-21 08:07:13 | 2021-07-21 08:07:13 |  20 | AnnouncementUpdated |               91 |          1 | NULL          | NULL        |
| 2613 |   12 |        NULL |         12 |             91 |          1 | 2613 | announcement | 2021-07-21 08:07:13 | 2021-07-21 08:07:13 |  20 | AnnouncementUpdated |               91 |          1 | NULL          | NULL        |
+------+------+-------------+------------+----------------+------------+------+--------------+---------------------+---------------------+-----+---------------------+------------------+------------+---------------+-------------+
5 rows in set (0.001 sec)
```